#!/usr/bin/env bash
# This script gets the consensus sequences for each case and checks for the presence of specific viruses.
# It concatenates the sequences with reference sequences for multiple sequencing alignments.
# $1 = input CSV file with sample names and paths
# $2 = path to the directory containing reference sequences

# list of viruses to process
virus_list=("FAdV" "IBDV" "IBV")

# extract unique case names from the CSV file generated by makecsv.sh
cut -f 1 -d ',' $1 | grep -v "SampleName" | cut -f 1 -d'-' | uniq > caselist.txt

# concatenate fasta files from each case into a single file
for case in $(cat caselist.txt);do cat ${case}-*.fasta > ${case}_consensus.fasta
	# loop through each virus in the list
	for virus in "${virus_list[@]}"; do
		# check if the virus is present in the concatenated file for the case
		if grep -q "$virus" ${case}_consensus.fasta; then
			# extract sequences for the virus and create a new virus only file
			awk -v v="$virus" '/^>/ {keep = ($0 ~ v)} keep' ${case}_consensus.fasta > "${case}_${virus}_only.fasta"
			#sed -i '/^>/! s/N//g' "${case}_${virus}_only.fasta" 
			# concatenate the extracted sequences with the reference sequences
			cat "${case}_${virus}_only.fasta" $2/${virus}_references.fasta > ${case}_${virus}_unaligned.fasta
			mafft --auto ${case}_${virus}_unaligned.fasta > ${case}_${virus}_msa.fasta
		else
			echo ">Novirusfromlist" > no_msa.fasta

		fi
	done
done