---
title: "Targeted amplicon sequencing results report"
author: "Generated using [ONT_targseq](https://github.com/dhineshp565/ONT_targseq) pipeline"
date: "`r Sys.Date()`"
output: html_document
params:
  csv: ""
  png: "*.png"
  krona: ""
  igv: ""
---

```{r setup, include=FALSE}
library(knitr)
library(ggplot2)
library(htmltools)
library(base64enc)
library(dplyr)
```


<br>
```{css, echo=FALSE}
.table caption {
    color: darkblue;
    font-weight: bold;
}
```

### Click dropdown to select sample {.tabset .tabset-dropdown}
```{r, results='asis',echo=FALSE}

samplelist <- read.csv(params$csv,header=FALSE,sep = ',')[-1,]
sampleid <- samplelist[,c(1)]

for (i in sampleid){
  cat("####", i, "\n")
  #use mappedreads.txt and display column 1 and 3
  mapped_reads <- (paste(i,"_mappedreads.txt",sep=""))
  df <- read.csv(mapped_reads,header = TRUE,sep=" ")
  mapped_stat<-df[,c(1,3)]
  colnames (mapped_stat) <- c("AMPLICON","MAPPED_READS")
  print(knitr::kable(mapped_stat,align="ll",caption = "No. of mapped reads. Generated using [minimap2](https://github.com/lh3/minimap2) and [samtools](https://github.com/samtools/samtools)"))

  cat('\n\n<!-- -->\n\n')
  
  # display abricate results
  abricate_ids <- (paste(i,"_abricate.csv",sep=""))
  abricate_raw <- read.csv(abricate_ids,header = TRUE,sep = "\t")
  abricate_final <- abricate_raw[,c(2,6,10,11,12,15)]
  colnames(abricate_final) <- c("SEQUENCE","GENE","%COVERAGE","%IDENTITY","DATABASE","REFERENCE") 
  print(knitr::kable(abricate_final,align = "llcccl",caption = "Consensus compared to reference database. Generated using [abricate](https://github.com/tseemann/abricate) and custom database")) 
   
  

  cat('\n\n<!-- -->\n\n')
  fasta_id <- (paste(i,"_consensus.fasta",sep=""))
  fasta_file <- readLines(fasta_id)
  # Display the content
  cat("```\n")
  cat(fasta_file, sep = "\n")
  cat("\n```\n")
  # Encode the file content as base64
  encoded_content <- base64encode(charToRaw(paste(fasta_file, collapse = "\n")))
  # Get file name from file path and remove extra underscores
  file_name <- gsub("_+", "_", basename(fasta_id))
  # Provide a download link for the encoded FASTA content with file name
  cat('<a href="data:text/plain;charset=utf-8;base64,', encoded_content, '" download="', file_name, '">Download FASTA file</a>')

  cat('\n\n<!-- -->\n\n')
  orf_id <- (paste(i,"_ORF.fasta",sep=""))
  orf_file <- readLines(orf_id)
  # Display the content
  cat("```\n")
  cat(orf_file, sep = "\n")
  cat("\n```\n")
  # Encode the file content as base64
  encoded_content <- base64encode(charToRaw(paste(orf_file, collapse = "\n")))
  # Get file name from file path and remove extra underscores
  file_name <- gsub("_+", "_", basename(orf_id))
  # Provide a download link for the encoded FASTA content with file name
  cat('<a href="data:text/plain;charset=utf-8;base64,', encoded_content, '" download="', file_name, '">Download ORF file</a>')
  
  cat('\n\n<!-- -->\n\n')

  # print orf length
  seqlen_id <- (paste(i,"_seqlengths.tsv",sep=""))
  seqlen_raw <- read.csv(seqlen_id, sep="\t", header=FALSE)
  colnames(seqlen_raw) <- c("SEQUENCE_NAME", "LENGTH")
  print(knitr::kable(seqlen_raw, align="ll", caption = "Sequence name and length"))

  cat('\n\n<!-- -->\n\n')
  # display blast results
  blast_ids <- (paste(i,"_report_blast.csv",sep=""))
  blast_raw <- read.csv(blast_ids,header = TRUE,sep = "\t")
  blast_final <- blast_raw[,c(1,3,4,5,6,10)]
  colnames(blast_final) <- c("QUERY_ID","LENGTH","%COVERAGE","%IDENTITY","E_VAL","STITLE")
  # Sort by %IDENTITY descending, then %COVERAGE descending
  blast_final <- blast_final %>%
  arrange(desc(`LENGTH`), desc(`%IDENTITY`), desc(`%COVERAGE`))
  print(knitr::kable(blast_final,align = "lccccl",caption = "Consensus compared to NCBI BLAST database. Generated using [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi)"))


  cat('\n\n<!-- -->\n\n')
 

    gene_list <- abricate_final$GENE
  
    case_id <- sub("-.*$", "", i) # Extract the case ID from the sample ID


  
    if (any(grepl("HEXON", gene_list, ignore.case = TRUE))) {
      tree_file <- paste0(case_id, "_FAdV_msa.png")
      if (file.exists(tree_file)) {
        cat("##### <span style='color:darkblue; font-weight:bold;'>FAdV Phylogenetic Tree</span>\n\n")
        cat("![](", tree_file, ")\n\n", sep = "")
        cat("*Multiple sequence alignment and phylogenetic tree generation using [MAFFT](https://mafft.cbrc.jp/alignment/server/index.html) and [IQ-Tree2](http://www.iqtree.org/). Visualization using [ggtree](https://guangchuangyu.github.io/software/ggtree/).\n\n")
     
      }
    }

    if (any(grepl("IBDV", gene_list, ignore.case = TRUE))) {
      tree_file <- paste0(case_id, "_IBDV_msa.png")
      if (file.exists(tree_file)) {
        cat("##### <span style='color:darkblue; font-weight:bold;'>IBDV Phylogenetic Tree</span>\n\n")
        cat("![](", tree_file, ")\n\n", sep = "")
        cat("*Multiple sequence alignment and phylogenetic tree generation using [MAFFT](https://mafft.cbrc.jp/alignment/server/index.html) and [IQ-Tree2](http://www.iqtree.org/). Visualization using [ggtree](https://guangchuangyu.github.io/software/ggtree/).\n\n")
      }
    }

   if (any(grepl("IBV", gene_list, ignore.case = TRUE))) {
      tree_file <- paste0(case_id, "_IBV_msa.png")
      if (file.exists(tree_file)) {
        cat("##### <span style='color:darkblue; font-weight:bold;'>IBV Phylogenetic Tree</span>\n\n")
        cat("![](", tree_file, ")\n\n", sep = "")
        cat("*Multiple sequence alignment and phylogenetic tree generation using [MAFFT](https://mafft.cbrc.jp/alignment/server/index.html) and [IQ-Tree2](http://www.iqtree.org/). Visualization using [ggtree](https://guangchuangyu.github.io/software/ggtree/).\n\n")
     
      }
    }
  cat('\n\n<!-- -->\n\n')
   
  }
```


## IGV report - Coverage and depth [IGV](https://software.broadinstitute.org/software/igv/)

```{r, echo=FALSE}
library("htmltools")
htmltools::tags$iframe(
  src = base64enc::dataURI(file= params$igv, mime="text/html; charset=UTF-8"),
  style="border:10; position:relative; top:0; left:; right::; bottom:; width:100%; height:800px"
)
```
<br>

<br>

## Taxonomic classification of raw reads using [Kraken2](https://ccb.jhu.edu/software/kraken2/)


```{r, echo=FALSE}
library("htmltools")
htmltools::tags$iframe(
  src = base64enc::dataURI(file= params$krona, mime="text/html; charset=UTF-8"),
  style="border:10; position:relative; top:0; left:; right::; bottom:; width:100%; height:800px"
)
```
<br>